# Data Cleaning

This tutorial will show you how to clean your data to be useful for building an ML model. It goes with a video at _______ which provides more context - I would recommend viewing that video to understand what we're doing here, and more importantly trying to do some of this yourself with another data set.

Here we are going to clean and prepare the data for a machine learning exercise to find patterns in survival on the Titanic.


## Imports

Let's import some libraries we'll need for this. All will become clear.

In [25]:
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer

## Get The Data

I've uploaded data to the files section (see left menu - folder icon - 'Data' folder).

In [26]:
data_df = pd.read_csv('https://raw.githubusercontent.com/idontwantyourspamthanks/youtube-data-cleaning-exercise/master/train.csv')

## The Data Set
Data shamelessly stolen from https://www.kaggle.com/competitions/titanic/data?select=train.csv - grab it and put it in files on the left menu in a new 'data' folder.

 - Survived = 1 for alive, 0 for dead
 - pclass = ticket class - 1 for 1st, 2 for 2nd, 3 for 3rd
 - sibsp = Siblings/spouses on Titanic
 - parch = Parents/children on Titanic
 - Ticket = ticket number
 - Embarked = Port of embarkation - C = Cherbourg, Q = Queenstown, S = Southampton

In [27]:
data_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

## Split The Data

We need to **split** the data into two parts. One to **train** the model on, and another to **validate** it. We need a set that is different to the data used to train the model to confirm that the model is not **overfitted** to the training data.

In [28]:
# Split the data randomly with an 80%/20% split between training and validation data
train_df, validation_df = train_test_split(data_df, test_size=0.2, random_state=42)

### The Training Data

This will become the data on which you train your model.

In [ ]:
train_df.head()

PassengerId  Survived  Pclass                                Name  \
331          332         0       1                 Partner, Mr. Austen   
733          734         0       2          Berriman, Mr. William John   
382          383         0       3                  Tikkanen, Mr. Juho   
704          705         0       3             Hansen, Mr. Henrik Juul   
813          814         0       3  Andersson, Miss. Ebba Iris Alfrida   

        Sex   Age  SibSp  Parch             Ticket     Fare Cabin Embarked  
331    male  45.5      0      0             113043  28.5000  C124        S  
733    male  23.0      0      0              28425  13.0000   NaN        S  
382    male  32.0      0      0  STON/O 2. 3101293   7.9250   NaN        S  
704    male  26.0      1      0             350025   7.8542   NaN        S  
813  female   6.0      4      2             347082  31.2750   NaN        S

### The Validation Data

You will then use this data to validate it, aiming to make predictions and match them against the Y field (survived).

In [ ]:
validation_df.head()

PassengerId  Survived  Pclass  \
709          710         1       3   
439          440         0       2   
840          841         0       3   
720          721         1       2   
39            40         1       3   

                                                  Name     Sex   Age  SibSp  \
709  Moubarek, Master. Halim Gonios ("William George")    male   NaN      1   
439             Kvillner, Mr. Johan Henrik Johannesson    male  31.0      0   
840                        Alhomaki, Mr. Ilmari Rudolf    male  20.0      0   
720                  Harper, Miss. Annie Jessie "Nina"  female   6.0      0   
39                         Nicola-Yarred, Miss. Jamila  female  14.0      1   

     Parch            Ticket     Fare Cabin Embarked  
709      1              2661  15.2458   NaN        C  
439      0        C.A. 18723  10.5000   NaN        S  
840      0  SOTON/O2 3101287   7.9250   NaN        S  
720      1            248727  33.0000   NaN        S  
39       0              2651  11.2417   NaN        C

## Features And Target

With our data split into training and validation, we next need to split each group into the features dataframe and the target dataframe. We are aiming to guess the target column 'Survived' based on the content of the other columns (features). With that in mind, we'll look to separate 'Survived' from the rest of the data.

In [29]:
train_X = train_df.drop('Survived', axis=1)
train_y = train_df['Survived']

valid_X = validation_df.drop('Survived', axis=1)
valid_y = validation_df['Survived']
# You can explore these using .head() if you wish
print("Done")

Done


# Ready To Clean

Right, we have the data we need, and it's split into training and validation, and into features and target, let's roll.

## Missing Values

Some data sets will miss values. It's inevitable. Consider a table of information about houses. A column with the size of bedroom 3 in sq ft might be missing if the house lacks a 3rd bedroom. We need to find ways to handle that.

### Which Columns have missing values?
We'll look at the data and see where there is missing information.

In [30]:
# Training data
missing_training_data = train_X.isnull().any()
missing_training_data_columns = missing_training_data[missing_training_data == True].index.tolist()
print(missing_training_data_columns)

['Age', 'Cabin', 'Embarked']


### So we know which columns have missing data
Age, Cabin and Embarked are the missing fields on the training data, and thankfully our validation data doesn't appear to be missing anything else. We're going to come back to Cabin as it's a more interesting case than first appears.

We don't look at the validation set - we assume it has similar characteristics but don't want **data leakage** - that is we don't want to start letting the validation set colour our build of the training model, otherwise the model might be overfit to the specifics of your validation and training sets rather than being able to account for real world data.

Now we can start thinking about what to do about those missing values.

In [31]:
missing_count_training = train_X.isnull().sum()
training_rows = len(train_X)

print (f'Training missing items:\n{missing_count_training[missing_count_training > 0]}\nfrom {training_rows} total rows')

Training missing items:
Age         140
Cabin       553
Embarked      2
dtype: int64
from 712 total rows


### Missing Data Discussion

So we have about 75% missing data in the Cabin column, and a little under 20% missing on the age column, and only a couple of missing items on Embarked. Does this data matter to our goal of figuring out patterns around who survived and who didn't?

One could argue that elderly people and small children might be more vulnerable than other groups, and that certain cabins were located in more dangerous locations than others.

## Age

We have most of the values here, which means there may be some value in trying to keep age, especially as it's very likely that it had an impact on survival. I hypothesise that kids would be less able to escape unaided, but would have priority for lifeboats, while the elderly would be less mobile and low priority.

One approach when we don't know is to impute the data - there are a number of approaches to this but for now we're going to use a simple mean approach, taking the mean of the column's values and filling the empty spaces with that value. We will also help the model out by adding a column to indicate that it was imputed. This gives the model the chance to determine that those items that are imputed might consider age to be less important.

In [32]:
imputer = SimpleImputer(strategy='mean')

train_X_age_imputed = train_X.copy(deep=True)
train_X_age_imputed['Age_Imputed'] = train_X_age_imputed['Age'].isnull().astype(int)
train_X_age_imputed['Age'] = imputer.fit_transform(train_X_age_imputed[['Age']])

## The imputer has been fitted, now it can be used to transform the valid set
valid_X_age_imputed = valid_X.copy(deep=True)
valid_X_age_imputed['Age_Imputed'] = valid_X_age_imputed['Age'].isnull().astype(int)
imputer.transform(valid_X_age_imputed[['Age']]) # [[]] ensures dataframe, not series, even when selecting a single column

train_X_age_imputed.head()

PassengerId  Pclass                                Name     Sex   Age  \
331          332       1                 Partner, Mr. Austen    male  45.5   
733          734       2          Berriman, Mr. William John    male  23.0   
382          383       3                  Tikkanen, Mr. Juho    male  32.0   
704          705       3             Hansen, Mr. Henrik Juul    male  26.0   
813          814       3  Andersson, Miss. Ebba Iris Alfrida  female   6.0   

     SibSp  Parch             Ticket     Fare Cabin Embarked  Age_Imputed  
331      0      0             113043  28.5000  C124        S            0  
733      0      0              28425  13.0000   NaN        S            0  
382      0      0  STON/O 2. 3101293   7.9250   NaN        S            0  
704      1      0             350025   7.8542   NaN        S            0  
813      4      2             347082  31.2750   NaN        S            0

## Embarked
So for this one we're only missing 2 values out of 712. It seems sensible to simply put a different category in for those so that we have C for Cherbourg, Q for Queenstown, S for Southampton and U for Unknown.

In [33]:
train_X_age_imputed['Embarked'] = train_X_age_imputed['Embarked'].fillna('U')
valid_X_age_imputed['Embarked'] = valid_X_age_imputed['Embarked'].fillna('U')

train_X_age_imputed.head()


PassengerId  Pclass                                Name     Sex   Age  \
331          332       1                 Partner, Mr. Austen    male  45.5   
733          734       2          Berriman, Mr. William John    male  23.0   
382          383       3                  Tikkanen, Mr. Juho    male  32.0   
704          705       3             Hansen, Mr. Henrik Juul    male  26.0   
813          814       3  Andersson, Miss. Ebba Iris Alfrida  female   6.0   

     SibSp  Parch             Ticket     Fare Cabin Embarked  Age_Imputed  
331      0      0             113043  28.5000  C124        S            0  
733      0      0              28425  13.0000   NaN        S            0  
382      0      0  STON/O 2. 3101293   7.9250   NaN        S            0  
704      1      0             350025   7.8542   NaN        S            0  
813      4      2             347082  31.2750   NaN        S            0

# Categorical Data

We've now got two bits of categorical data. Sex and Embarked. They have slightly different approaches based on the number of unique values. We can't do maths on text, so we need to turn this data into numbers, but how?

With 4 values, if we numbered the embarkation 1, 2, 3, 4 it might indicate that 4 was more valuable than 1, and thus the model could make mistakes based on assumptions of linearity. Instead we wish to indicate true or false whether each of the possible conditions is true. We will create a column for each of these. We would not do this where there were 50 categories, as that would make the table too big to realistically work on.

Biological sex being either male or female we could represent with 1 and 2, as having only two values means there's not a scale to consider. However, to play it safe we'll use the same method of encoding here as for Embarkation.

The method is called **One Hot Encoding**

In [40]:
encoder = OneHotEncoder(sparse_output=False, drop='first')

# Fit on training and transform both
encoded_train = encoder.fit_transform(train_X_age_imputed[['Sex', 'Embarked']])
encoded_val = encoder.transform(valid_X_age_imputed[['Sex', 'Embarked']])
# These are now arrays, not dataframes

# Get new column names after encoding
encoded_columns = encoder.get_feature_names_out(['Sex', 'Embarked'])

# Turn the new columns (currently as arrays) into DataFrames
encoded_train_df = pd.DataFrame(encoded_train, columns=encoded_columns, index=train_X_age_imputed.index)
encoded_val_df = pd.DataFrame(encoded_val, columns=encoded_columns, index=valid_X_age_imputed.index)

# Join these new columns to the original DataFrames
oh_train_df = pd.concat([train_X_age_imputed, encoded_train_df], axis=1).drop(['Sex', 'Embarked'], axis=1)
oh_valid_df = pd.concat([valid_X_age_imputed, encoded_val_df], axis=1).drop(['Sex', 'Embarked'], axis=1)


## Cabin

The cabin column is interesting. The data contains values like C85, E392 and so on. The letter indicates which floor, and the numbers indicate position along the boat. My gut feeling is that the floor is relevant. I'd like a bit more data to work with, as 75% is missing, but we'll use what we've got as it's the only column where the categories are clearly sequential, plus it lets us do some more substantial Python.

In [41]:
# Map the unique values to numerical codes
deck_mapping = {'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E': 5, 'F': 6, 'G': 7, 'T': 8}

oh_train_df['Deck'] = oh_train_df['Cabin'].str[0]
oh_train_df['Deck'] = oh_train_df['Deck'].map(deck_mapping)
oh_train_df['Deck'] = oh_train_df['Deck'].fillna(0)
oh_train_df['Deck_Imputed'] = (oh_train_df['Deck'] == 0).astype(int)

oh_valid_df['Deck'] = oh_valid_df['Cabin'].str[0]
oh_valid_df['Deck'] = oh_valid_df['Deck'].map(deck_mapping)
oh_valid_df['Deck'] = oh_valid_df['Deck'].fillna(0)
oh_valid_df['Deck_Imputed'] = (oh_valid_df['Deck'] == 0).astype(int)

oh_train_df.drop('Cabin', axis=1, inplace=True)
oh_valid_df.drop('Cabin', axis=1, inplace=True)


# Irrelevant Data

So we can't really do anything useful with name or ticket - they don't indicate anything related to survivability and they're very difficult if not impossible to turn into quantitative data. PassengerId doesn't tell us anything useful either. Let's remove them.

In [42]:
oh_train_df.drop(['PassengerId', 'Name', 'Ticket'], axis=1, inplace=True)
oh_valid_df.drop(['PassengerId', 'Name', 'Ticket'], axis=1, inplace=True)

oh_train_df

Pclass        Age  SibSp  Parch      Fare  Age_Imputed  Sex_male  \
331       1  45.500000      0      0   28.5000            0       1.0   
733       2  23.000000      0      0   13.0000            0       1.0   
382       3  32.000000      0      0    7.9250            0       1.0   
704       3  26.000000      1      0    7.8542            0       1.0   
813       3   6.000000      4      2   31.2750            0       0.0   
..      ...        ...    ...    ...       ...          ...       ...   
106       3  21.000000      0      0    7.6500            0       0.0   
270       1  29.498846      0      0   31.0000            1       1.0   
860       3  41.000000      2      0   14.1083            0       1.0   
435       1  14.000000      1      2  120.0000            0       0.0   
102       1  21.000000      0      1   77.2875            0       1.0   

     Embarked_Q  Embarked_S  Embarked_U  Deck  Deck_Imputed  
331         0.0         1.0         0.0   3.0             0  
733         0.0         1.0         0.0   0.0             1  
382         0.0         1.0         0.0   0.0             1  
704         0.0         1.0         0.0   0.0             1  
813         0.0         1.0         0.0   0.0             1  
..          ...         ...         ...   ...           ...  
106         0.0         1.0         0.0   0.0             1  
270         0.0         1.0         0.0   0.0             1  
860         0.0         1.0         0.0   0.0             1  
435         0.0         1.0         0.0   2.0             0  
102         0.0         1.0         0.0   4.0             0  

[712 rows x 12 columns]

## Feature engineering

We can get family size from adding the SibSp and Parch values, and of course adding the individual. Might this be a factor in survival? Let's see.

In [15]:
oh_train_df['FamilySize'] = oh_train_df['SibSp'] + oh_train_df['Parch'] + 1
oh_valid_df['FamilySize'] = oh_valid_df['SibSp'] + oh_valid_df['Parch'] + 1
oh_valid_df

Pclass   Age  SibSp  Parch     Fare Cabin  Age_Imputed  Sex_male  \
709       3   NaN      1      1  15.2458   NaN            1       1.0   
439       2  31.0      0      0  10.5000   NaN            0       1.0   
840       3  20.0      0      0   7.9250   NaN            0       1.0   
720       2   6.0      0      1  33.0000   NaN            0       0.0   
39        3  14.0      1      0  11.2417   NaN            0       0.0   
..      ...   ...    ...    ...      ...   ...          ...       ...   
433       3  17.0      0      0   7.1250   NaN            0       1.0   
773       3   NaN      0      0   7.2250   NaN            1       1.0   
25        3  38.0      1      5  31.3875   NaN            0       0.0   
84        2  17.0      0      0  10.5000   NaN            0       0.0   
10        3   4.0      1      1  16.7000    G6            0       0.0   

     Embarked_Q  Embarked_S  Embarked_U  FamilySize  
709         0.0         0.0         0.0           3  
439         0.0         1.0         0.0           1  
840         0.0         1.0         0.0           1  
720         0.0         1.0         0.0           2  
39          0.0         0.0         0.0           2  
..          ...         ...         ...         ...  
433         0.0         1.0         0.0           1  
773         0.0         0.0         0.0           1  
25          0.0         1.0         0.0           7  
84          0.0         1.0         0.0           1  
10          0.0         1.0         0.0           3  

[179 rows x 12 columns]

## Scaling Features

Finally let's scale the features. One column having ranges from 0 to 20000 vs another with range 1-40 for instance might give undue influence to one column purely on account of its size.

In [19]:
scaler = StandardScaler()
final_X_train = pd.DataFrame(scaler.fit_transform(oh_train_df), columns=oh_train_df.columns)
final_X_valid = pd.DataFrame(scaler.transform(oh_valid_df), columns=oh_valid_df.columns)

# Final Data State

Let's have a look at the final state of the training data.

In [20]:
final_X_train

Pclass           Age     SibSp     Parch      Fare  Age_Imputed  \
0   -1.614136  1.232263e+00 -0.470722 -0.479342 -0.078684    -0.494727   
1   -0.400551 -5.004820e-01 -0.470722 -0.479342 -0.377145    -0.494727   
2    0.813034  1.926161e-01 -0.470722 -0.479342 -0.474867    -0.494727   
3    0.813034 -2.694493e-01  0.379923 -0.479342 -0.476230    -0.494727   
4    0.813034 -1.809667e+00  2.931860  2.048742 -0.025249    -0.494727   
..        ...           ...       ...       ...       ...          ...   
707  0.813034 -6.545038e-01 -0.470722 -0.479342 -0.480162    -0.494727   
708 -1.614136  2.735977e-16 -0.470722 -0.479342 -0.030545     2.021315   
709  0.813034  8.857142e-01  1.230569 -0.479342 -0.355804    -0.494727   
710 -1.614136 -1.193580e+00  0.379923  2.048742  1.683201    -0.494727   
711 -1.614136 -6.545038e-01 -0.470722  0.784700  0.860748    -0.494727   

     Sex_male  Embarked_Q  Embarked_S  Embarked_U  FamilySize      Deck  
0    0.724310   -0.303355    0.596817   -0.053074   -0.554666  1.423001  
1    0.724310   -0.303355    0.596817   -0.053074   -0.554666 -0.480274  
2    0.724310   -0.303355    0.596817   -0.053074   -0.554666 -0.480274  
3    0.724310   -0.303355    0.596817   -0.053074    0.040096 -0.480274  
4   -1.380624   -0.303355    0.596817   -0.053074    3.013909 -0.480274  
..        ...         ...         ...         ...         ...       ...  
707 -1.380624   -0.303355    0.596817   -0.053074   -0.554666 -0.480274  
708  0.724310   -0.303355    0.596817   -0.053074   -0.554666 -0.480274  
709  0.724310   -0.303355    0.596817   -0.053074    0.634859 -0.480274  
710 -1.380624   -0.303355    0.596817   -0.053074    1.229621  0.788576  
711  0.724310   -0.303355    0.596817   -0.053074    0.040096  2.057427  

[712 rows x 12 columns]